src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -71,99 +83,308 @@ public long BytesWritten
+        /// In the case of Stream, this writes the data to the stream and flushes it.
+        /// </remarks>
+        /// <remarks>
+        /// The <see cref="Utf8JsonWriter"/> instance cannot be re-used after disposing.
+        /// </remarks>
+        public async ValueTask DisposeAsync()
+        {
+            if (_output == null)
+            {
+                return;
+            }
+
+            await FlushAsync().ConfigureAwait(false);
+            ResetHelper();
+
+            _stream = null;
+            _arrayBufferWriter = null;
+            _output = null;
+        }
+#endif
+
+        /// <summary>
+        /// Asynchronously commits the JSON text written so far which makes it visible to the output destination.
+        /// </summary>
+        /// <remarks>
+        /// In the case of IBufferWriter, this advances the underlying <see cref="IBufferWriter{Byte}" /> based on what has been written so far.
+        /// In the case of Stream, this writes the data to the stream and flushes it asynchronously, while monitoring cancellation requests.
+        /// </remarks>
+        public async Task FlushAsync(CancellationToken cancellationToken = default)
+        {
+            FlushHelper();
+
+            if (_stream != null)
+            {
+                Debug.Assert(_arrayBufferWriter != null);
+                Debug.Assert(BytesPending == _arrayBufferWriter.WrittenCount);
+                if (BytesPending != 0)
+                {
+#if BUILDING_INBOX_LIBRARY
+                    await _stream.WriteAsync(_arrayBufferWriter.WrittenMemory, cancellationToken).ConfigureAwait(false);
+#else
+                    Debug.Assert(_arrayBufferWriter.WrittenMemory.Length == _arrayBufferWriter.WrittenCount);
+                    await _stream.WriteAsync(_arrayBufferWriter.WrittenMemory.ToArray(), 0, _arrayBufferWriter.WrittenCount, cancellationToken).ConfigureAwait(false);
+#endif
+                    _arrayBufferWriter.Clear();
+                }
+                await _stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            _memory = default;
+            BytesCommitted += BytesPending;
+            BytesPending = 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void FlushHelperStream()
+        {
+            Debug.Assert(_arrayBufferWriter != null);
+            Debug.Assert(BytesPending == _arrayBufferWriter.WrittenCount);
+            if (BytesPending != 0)
+            {
+#if BUILDING_INBOX_LIBRARY
+                _stream.Write(_arrayBufferWriter.WrittenSpan);
+#else
+                Debug.Assert(_arrayBufferWriter.WrittenSpan.Length == _arrayBufferWriter.WrittenCount);
+                _stream.Write(_arrayBufferWriter.WrittenSpan.ToArray(), 0, _arrayBufferWriter.WrittenCount);
+#endif
+                _arrayBufferWriter.Clear();
```

Comment:
We end up clearing the writer, which means zero'ing all bytes written, every time we flush.  That seems unnecessary.  Shouldn't we be able to just reset the position to 0 and not clear at all?  When we're done using the buffer writer we'll just drop it and let the GC clean up.

Reply:
Agreed.

Reply:
We don't have a "reset" method on the ArrayBufferWriter that just resets the position 0. Should we make the `WrittenCount` property settable?

Reply:
> We don't have a "reset" method on the ArrayBufferWriter that just resets the position 0. Should we make the WrittenCount property settable, or just change Clear() to Reset() which doesn't clear?

That would be my preference, with it also renamed to `Position` :wink: